type_info.go

```diff
@@ -0,0 +1,240 @@
+package duckdb
+
+/*
+#include <duckdb.h>
+*/
+import "C"
+
+import (
+	"reflect"
+	"runtime"
+	"unsafe"
+)
+
+type structEntry struct {
+	TypeInfo
+	name string
+}
+
+// StructEntry is an interface to provide STRUCT entry information.
+type StructEntry interface {
+	// Info returns a STRUCT entry's type information.
+	Info() TypeInfo
+	// Name returns a STRUCT entry's name.
+	Name() string
+}
+
+// NewStructEntry returns a STRUCT entry.
+func NewStructEntry(info TypeInfo, name string) (StructEntry, error) {
+	if name == "" {
+		return nil, getError(errAPI, errEmptyName)
+	}
+
+	return &structEntry{
+		TypeInfo: info,
+		name:     name,
+	}, nil
+}
+
+// Info returns a STRUCT entry's type information.
+func (entry *structEntry) Info() TypeInfo {
+	return entry.TypeInfo
+}
+
+// Name returns a STRUCT entry's name.
+func (entry *structEntry) Name() string {
+	return entry.name
+}
+
+type baseTypeInfo struct {
+	Type
+	structEntries []StructEntry
+	decimalWidth  uint8
+	decimalScale  uint8
+}
+
+type vectorTypeInfo struct {
+	baseTypeInfo
+	dict map[string]uint32
+}
+
+type typeInfo struct {
+	baseTypeInfo
+	childTypes []TypeInfo
+	enumNames  []string
+}
+
+// TypeInfo is an interface for DuckDB types.
+type TypeInfo interface {
+	logicalType() C.duckdb_logical_type
+}
```

Comment:
I've turned `TypeInfo` into an `interface`. My only concern is that developers might run into `nil` pointer dereferences when passing uninitialized interfaces to functions if we no longer return an error instead. Is this fine for idiomatic Go? Or should each function that accepts an interface also contain a `nil` check and return an error?

The same question applies to `type StructEntry interface`.

Reply:
Hmm, you raise a good point. I think the best way it indeed to either return an error, or default to `any` (which is just a type without any typeinfo right?).

 for `type StructEntry interface` I think it should just be a `struct{info Typeinfo, name string}` avoiding this whole issue (with a nil typeinfo still being `any`).
 
 Sorry if I'm a bit opinionated, but I think this is a very nice API :)

Reply:
No worries; I appreciate the discussion and input.
I'll then add `nil` checks and return an appropriate error.

W.r.t. `StructEntry`, I prefer using the interface, as we stay flexible with the internals and can add error checks during creation, like the empty name check.